ydb/core/audit/audit_log_impl.cpp

```diff
@@ -112,19 +110,34 @@ TString GetJsonLogCompatibleLog(const TEvAuditLog::TEvWriteAuditLog::TPtr& ev) {
     return ss.Str();
 }
 
-TString GetTxtLog(const TEvAuditLog::TEvWriteAuditLog::TPtr& ev) {
-    const auto* msg = ev->Get();
+TString GetTxtLog(TInstant time, const TAuditLogParts& parts) {
     TStringStream ss;
-    ss << msg->Time << ": ";
-    for (auto it = msg->Parts.begin(); it != msg->Parts.end(); it++) {
-        if (it != msg->Parts.begin())
+    ss << time << ": ";
+    for (auto it = parts.begin(); it != parts.end(); it++) {
+        if (it != parts.begin())
             ss << ", ";
         ss << it->first << "=" << it->second;
     }
     ss << Endl;
     return ss.Str();
 }
 
+// Array of functions for converting TEvAuditLog::TEvWriteAuditLog events to a string.
+// Indexing in the array occurs by the value of the NKikimrConfig::TAuditConfig::EFormat enumeration.
+// For each new format, we need to register the audit event conversion function.
+// The size of AuditLogItemBuilders must be equal to the maximum value of the NKikimrConfig::TAuditConfig::EFormat enumeration.
+static std::vector<TAuditLogItemBuilder> AuditLogItemBuilders = { GetJsonLog, GetTxtLog, GetJsonLogCompatibleLog, nullptr };
+
+// numbering enumeration starts from one
+static constexpr size_t DefaultAuditLogItemBuilder = static_cast<size_t>(NKikimrConfig::TAuditConfig::JSON) - 1;
```

Comment:
There are several constants [defined for a protobuf enum](https://protobuf.dev/reference/cpp/cpp-generated/#enum) in the generated code: MIN, MAX, ARRAYSIZE (also IsValid() and conversion routines).

So for offsetting an index I would recommend to use `TAuditConfig::EFormat_MIN` instead of literal `1`.

Or may be even better -- just add to the `AuditLogItemBuilders` a dummy nullptr at the index 0 and there will be a direct mapping between enum values and array indexes, no offsetting will be needed.